src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseInviteService.cs

```diff
@@ -0,0 +1,231 @@
+using static GovUk.Education.ExploreEducationStatistics.Admin.Validators.ValidationErrorMessages;
+using static GovUk.Education.ExploreEducationStatistics.Content.Model.ReleaseRole;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Services
+{
+    public class ReleaseInviteService : IReleaseInviteService
+    {
+        private readonly ContentDbContext _contentDbContext;
+        private readonly IPersistenceHelper<ContentDbContext> _contentPersistenceHelper;
+        private readonly IUserRepository _userRepository;
+        private readonly IUserService _userService;
+        private readonly IUserInviteRepository _userInviteRepository;
+        private readonly IUserReleaseInviteRepository _userReleaseInviteRepository;
+        private readonly IUserReleaseRoleRepository _userReleaseRoleRepository;
+        private readonly IConfiguration _configuration;
+        private readonly IEmailService _emailService;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+
+        public ReleaseInviteService(ContentDbContext contentDbContext,
+            IPersistenceHelper<ContentDbContext> contentPersistenceHelper,
+            IUserRepository userRepository,
+            IUserService userService,
+            IUserInviteRepository userInviteRepository,
+            IUserReleaseInviteRepository userReleaseInviteRepository,
+            IUserReleaseRoleRepository userReleaseRoleRepository,
+            IConfiguration configuration,
+            IEmailService emailService,
+            IHttpContextAccessor httpContextAccessor)
+        {
+            _contentDbContext = contentDbContext;
+            _contentPersistenceHelper = contentPersistenceHelper;
+            _userRepository = userRepository;
+            _userService = userService;
+            _userInviteRepository = userInviteRepository;
+            _userReleaseInviteRepository = userReleaseInviteRepository;
+            _userReleaseRoleRepository = userReleaseRoleRepository;
+            _configuration = configuration;
+            _emailService = emailService;
+            _httpContextAccessor = httpContextAccessor;
+        }
+
+        public async Task<Either<ActionResult, Unit>> InviteContributor(string email, Guid publicationId,
+            List<Guid> releaseIds)
+        {
+            return await _contentPersistenceHelper
+                .CheckEntityExists<Publication>(publicationId, query => query
+                    .Include(p => p.Releases))
+                .OnSuccessDo(
+                    async publication => await _userService.CheckCanUpdateReleaseRole(publication, Contributor))
+                .OnSuccessDo(() => ValidateEmail(email))
+                .OnSuccess(publication => ValidateReleaseIds(publication, releaseIds))
+                .OnSuccess(async publication =>
+                {
+                    email = email.Trim();
+                    releaseIds = releaseIds.Distinct().ToList();
```

Comment:
I've added these two lines - I wanted to highlight them in case others think I should do this a different way.

I've trimmed the email string to avoid duplicate entries for the same email being possible due to whitespace.
I've made releaseIds distinct, because I realised that maybe issues could occur if the same releaseId was provided twice in a request. Not sure that it actually would, but this removes all possibility.

Reply:
Agree about the release id's. Could it potentially be hoisted any higher in this method as we also check all the release id's belong to the publication?

For the email address, it's a shame the validator doesn't prevent against this. Looking at the code of it, I don't think it does. If this input can come from a user then I'd also check we are doing 'ToLower' with anywhere that it's being used. I think that's probably already the case with all the repository classes though.

Reply:
I've moved the duplicate releaseIds check into ValidateReleaseIds, where it throws an exception if there are duplicates.

I checked and email.ToLower() is called everywhere except where the email is sent.

Reply:
I've moved the releaseIds distinct check into ValidateReleaseIds method.

Discussed emails with Ben on Slack a little. Feels like we need some kind of central validate/clean email class that all email arguments get pushed through - and then all the various ad hoc email validation and cleaning / ToLower() can be removed. I think this can be done as part of the work for EES-2989